runtime/src/main/java/org/corfudb/protocols/wireprotocol/IMetadata.java

```diff
@@ -78,10 +79,30 @@ default Long getRank() {
      * @param rank The rank of this append.
      */
     default void setRank(long rank) {
+        if (rank>=0 && getRank()<0) {
+            throw new IllegalStateException("Cannot set rank when the overwrite is permitted");
```

Comment:
Actually in append(...) you don't hold the lock, you hold it on writeRecord(...) that is called from append, but this is solvable.

I believe this was the reason I decided not to read the old rank here - I didn't want to change your locks for such rare situation. The disadvantage was that the responsibility to force overwrite was moving to the client.

I don't believe if the write lock is extended to append from write, something bad will happen.